File: appliance-root/opt/napp/www/js/pki.js
jQuery(function ($) {
  function set_state() {
    $.ajax({
      url: "/api/json/provisioned_state",
      type: "GET",
      dataType: "json",
      success: function (data, status, request) {
        $('#pki-key-ready').removeClass().addClass(
          data.details.key.exists ? 'exists' : 'missing'
        );
        $('#pki-ca-ready').removeClass().addClass(
          data.details.ca.exists ? 'exists' : 'missing'
        );
        if(data.details.crl.file)
          $('#pki-crl-ready').removeClass().addClass(
            data.details.crl.exists ? 'exists' : 'missing'
          );
        else
          $('#pki-crl-ready').removeClass().addClass('na');
        if(data.pki) {
          /* Allowed to go on */
          $("#fetch").text("Reconfigure");
          $('#fetch').unbind("click");
          $('#fetch').bind("click", reconfigure);
          if(!data.provisioned) $('#provision-step').show();
          else $('#dash-step').show();
        } else {
          $('#fetch').unbind("click");
          $('#fetch').bind("click", step);
        }
      },
      error: function (request, status, error) {
        alert('Error checking broker state');
      }
    });
  }
  function reconfigure() {
    fetchkey();
    fetchCA('ca');
    fetchCA('crl');
    step();
  }
  function step() {
    $.ajax({
      url: "/api/json/provisioned_state",
      type: "GET",
      dataType: "json",
      success: function (data, status, request) {
        // step 1
        if(!data.details.key.exists) return fetchkey();
        else $('#pki-key-ready').removeClass().addClass('exists');
        // step 2
        if(!data.details.ca.exists) return fetchCA('ca');
        else $('#pki-ca-ready').removeClass().addClass('exists');
        // step 3 (optional)
        if(data.details.crl.file) {
          if(!data.details.crl.exists) return fetchCA('crl');
          else $('#pki-crl-ready').removeClass().addClass('exists');
        }
        else $('#pki-crl-ready').removeClass().addClass('na');
        if(data.pki) {
          /* Allowed to go on */
          $("#fetch").text("Reconfigure");
          if(!data.provisioned) $('#provision-step').show();
          else $('#dash-step').show();
        }
      },
      error: function(request, status, error) {
        alert('Could not fetch status of PKI on the broker.');
      }
    });
  }
  function fetchkey() {
    $('#pki-key-ready').removeClass().addClass('progress');
    $.ajax({
      url: "/api/json/keygen",
      type: "GET",
      dataType: "json",
      success: function (data,status,request) {
        if(data.status == "success") return step();
        $('#pki-key-ready').removeClass().addClass('missing');
        alert('Failed to generate new private key: ' + data.error);
      },
      error: function(request, status, error) {
        $('#pki-key-ready').removeClass().addClass('missing');
        alert('Failed to generate new private key');
      }
    });
  }
  function fetchCA(type) {
    var $status = $('#pki-'+type+'-ready')
    $status.removeClass().addClass('progress');
    $.ajax({
      url: "/api/json/cafetch?type=" + type,
      type: "POST",
      data: {
        pki_url: $('#pki_url').val(),
        type: type
      },
      dataType: "json",
      success: function (data,status,request) {
        if(data.status == "success") return step();
        $status.removeClass().addClass('missing');
        alert('Failed to fetch ' + type + ': ' + data.error);
      },
      error: function(request, status, error) {
        $status.removeClass().addClass('missing');
        alert('Failed to fetch ' + type);
      }
    });
  }
  
  var urls = $('#pki');
  if(document.location.hash == "#inside") {
    urls.show();
    set_state();
  }
  else {
    set_state();
    step();
  }
});

File: appliance-root/opt/napp/www/pki.lua
http:write([=[<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <link href="/c/s.css" rel="stylesheet" media="screen" type="text/css" />
    <link rel="shortcut icon" type="image/ico" href="/favicon.ico" />
  </head>
  <body class="enterprise">
    <div id="masthead">
      <div id="header">
        <span class="app-logo"><a href="http://circonus.com/" title="Circonus Home">Circonus | Organization-wide Monitoring</a></span>
      </div>
    </div>
    <div class="page">
    <div id="page-content" class="clear">
      <div class="content-col">
      <img class="logo" src="/i/content/circonus-enterprise-icon.png" />
      <p class="intro">
                Please check to make sure all of the following security settings are correct.
      </p>
      <br style="clear:both"/>
        <h2 id="provision-head" class="form-head">Configure Security</h2>
      <br style="clear:both"/>
        <form id="login_form" method="POST">
          <fieldset id="pki" class="first_step">
            <legend><em>1.</em> Security Information</legend>
            <div id="circonus-pki">
              <label for="circonus_url">Circonus Base URL</label>
              <input id="circonus_url" type="text" name="circonus_url" value="]=])
    http:write(circonus_url())
    http:write([=[" />
              <button id="fetch" type="button">Configure</button>
            </div>
          </fieldset>
          <div id="pki-progress">
            <ul>
              <li id="pki-key-ready" class="pending"><div><span class="state"><span class="lock-label">Private Key</span></span></div></li>
              <li id="pki-ca-ready"  class="pending"><div><span class="state"><span class="lock-label">Certificate Authority</span></span></div></li>
              <li id="pki-crl-ready"  class="pending"><div><span class="state"><span class="lock-label">Certificate Revocation List</span></span></div></li>
            </ul>
          </div>
        </form>
      </div>
      <div id="provision-step">
        <button onclick="javascript:document.location='/provision]=])
if inside() then
  http:write("#inside")
end
http:write([=[';">Provision this Broker</button>
      </div>
      <div id="dash-step">
        <button onclick="javascript:document.location='/dash';">Go to the Broker's Dashboard &raquo;</button>
      </div>
    </div>
    </div>
    <script src="/js/jquery.min.js" type="text/javascript"></script>
    <script src="/js/pki.js" type="text/javascript"></script>
  </body>
</html>
]=])
